File: flutter/packages/hatake_material/test/app_tabs_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hatake_material/hatake_material.dart';

/// 画面を**並べて開く**（`app.navigation: tabs`）。
///
/// 業務システムによって作法が違うので、どちらかに決め打ちしない。定義が既定を言い、
/// アプリ側（`HatakeApp(navigation:)`）が上書きする。
///
/// ここで守るのは5つ。**既定は遷移**（書かなければいままでの動き）・**タブの中身は
/// 残る**（行き来しても検索条件が消えない＝それがタブの値打ち）・**同じ画面は2枚
/// 開かない**・**入力する画面を閉じるときは聞く**・**アプリの上書きが定義より強い**。
class _Repo implements Repository {
  @override
  Future<PageResult> search(RepositoryQuery query) async => PageResult(
        items: [
          {'id': 1, 'code': 'C001', 'note': '${query.filters['code'] ?? ''}'},
        ],
        totalCount: 1,
      );
  @override
  Future<DataRecord?> findByKey(Object key) async => {'id': key, 'code': 'C001'};
  @override
  Future<DataRecord> create(DataRecord data) async => data;
  @override
  Future<DataRecord> update(Object key, DataRecord data) async => data;
  @override
  Future<void> delete(Object key) async {}
}

/// 顧客一覧（行から詳細へ2通りの遷移）＋商品マスタ（入力できる画面）。
AppDefinition _app({String navigation = AppNavigation.single}) => AppDefinition(
      id: 'shop',
      title: 'ショップ',
      home: 'customers',
      navigation: navigation,
      menu: const [
        MenuItem(id: 'customers', label: '顧客', page: 'customers'),
        MenuItem(id: 'products', label: '商品', page: 'products'),
      ],
      pages: const [
        SearchPageDefinition(
          id: 'customers',
          title: '顧客一覧',
          repository: 'repo',
          search: SearchDefinition(
            filters: [FilterDefinition(field: 'code', label: 'コード')],
          ),
          table: TableDefinition(
            rowActions: ['open', 'openTab'],
            columns: [ColumnDefinition(field: 'code', label: 'コード')],
          ),
          actions: [
            // 既定（same）＝同じタブの中で進む。
            ActionDefinition(
              id: 'open',
              type: 'navigate',
              label: '詳細',
              config: {
                'page': 'customer_detail',
                'params': {'id': r'$row.id'},
              },
            ),
            // `open: tab` ＝一覧を残したまま別のタブで開く。
            ActionDefinition(
              id: 'openTab',
              type: 'navigate',
              label: '別タブ',
              open: ActionOpen.tab,
              config: {
                'page': 'customer_detail',
                'params': {'id': r'$row.id'},
              },
            ),
          ],
        ),
        DetailPageDefinition(
          id: 'customer_detail',
          title: '顧客詳細',
          repository: 'repo',
          form: FormDefinition(
            sections: [
              SectionDefinition(
                fields: [FieldDefinition(field: 'code', label: 'コード')],
              ),
            ],
          ),
        ),
        CrudPageDefinition(
          id: 'products',
          title: '商品マスタ',
          repository: 'repo',
          table: TableDefinition(
            columns: [ColumnDefinition(field: 'code', label: 'コード')],
          ),
          form: FormDefinition(),
        ),
      ],
    );

Widget _harness(AppDefinition app, {String? navigation}) => MaterialApp(
      home: HatakeScope(
        repositories: RepositoryRegistry({'repo': _Repo()}),
        renderer: const MaterialRenderer(),
        child: HatakeApp(app: app, syncUrl: false, navigation: navigation),
      ),
    );

/// メニューから選ぶ（幅の広い画面なので、左の一覧をそのまま押せる）。
Future<void> _pickMenu(WidgetTester tester, String label) async {
  await tester.tap(find.text(label).last);
  await tester.pumpAndSettle();
}

void main() {
  testWidgets('既定は遷移（タブの列は出ない・メニューで入れ替わる）', (tester) async {
    await tester.pumpWidget(_harness(_app()));
    await tester.pumpAndSettle();

    expect(find.byType(InputChip), findsNothing);
    await _pickMenu(tester, '商品');
    expect(find.text('商品マスタ'), findsWidgets);
    expect(find.text('顧客一覧'), findsNothing); // 入れ替わった
  });

  testWidgets('tabs ならメニューで選ぶと並ぶ（中身は残る）', (tester) async {
    await tester.pumpWidget(_harness(_app(navigation: AppNavigation.tabs)));
    await tester.pumpAndSettle();

    // 1枚目に検索条件を入れる（タブを行き来しても消えないこと）。
    await tester.enterText(
      find.byKey(const Key('hatake.filter.code')),
      'C001',
    );
    await tester.pumpAndSettle();

    await _pickMenu(tester, '商品');
    expect(find.byType(InputChip), findsNWidgets(2));

    // 1枚目に戻ると、入れた条件がそのまま在る（画面を作り直していない）。
    await tester.tap(find.byType(InputChip).first);
    await tester.pumpAndSettle();
    expect(
      tester
          .widget<TextField>(find.byKey(const Key('hatake.filter.code')))
          .controller
          ?.text,
      'C001',
    );
  });

  testWidgets('同じ画面をもう一度選んでも増えない（前に出るだけ）', (tester) async {
    await tester.pumpWidget(_harness(_app(navigation: AppNavigation.tabs)));
    await tester.pumpAndSettle();
    await _pickMenu(tester, '商品');
    expect(find.byType(InputChip), findsNWidgets(2));

    await _pickMenu(tester, '顧客');
    await _pickMenu(tester, '商品');
    expect(find.byType(InputChip), findsNWidgets(2));
  });

  testWidgets('open: tab の遷移は別のタブ、既定は同じタブの中で進む', (tester) async {
    await tester.pumpWidget(_harness(_app(navigation: AppNavigation.tabs)));
    await tester.pumpAndSettle();

    // 既定（same）＝タブは増えず、戻る口が出る。
    await tester.tap(find.byKey(const Key('hatake.rowaction.open.1')));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNWidgets(1));
    expect(find.byKey(const Key('hatake.app.back')), findsOneWidget);

    await tester.tap(find.byKey(const Key('hatake.app.back')));
    await tester.pumpAndSettle();

    // `open: tab` ＝一覧を残したまま2枚目が開く。
    await tester.tap(find.byKey(const Key('hatake.rowaction.openTab.1')));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNWidgets(2));
    expect(find.text('顧客詳細'), findsWidgets);
  });

  testWidgets('入力する画面を閉じるときは聞く（照会は聞かない）', (tester) async {
    await tester.pumpWidget(_harness(_app(navigation: AppNavigation.tabs)));
    await tester.pumpAndSettle();
    await _pickMenu(tester, '商品'); // CRUD＝入力できる画面

    // 閉じる口を押すと確認が出る。やめれば閉じない。
    await tester.tap(find.byKey(const Key('hatake.app.tab.1.close')));
    await tester.pumpAndSettle();
    expect(find.byKey(const Key('hatake.app.tab.confirmClose')), findsOneWidget);
    await tester.tap(find.text('やめる'));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNWidgets(2));

    // もう一度押して、閉じる。
    await tester.tap(find.byKey(const Key('hatake.app.tab.1.close')));
    await tester.pumpAndSettle();
    await tester.tap(find.byKey(const Key('hatake.app.tab.confirmClose.ok')));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNWidgets(1));
  });

  testWidgets('最後の1枚には閉じる口を出さない（画面が無くなる）', (tester) async {
    await tester.pumpWidget(_harness(_app(navigation: AppNavigation.tabs)));
    await tester.pumpAndSettle();

    expect(find.byType(InputChip), findsNWidgets(1));
    expect(find.byKey(const Key('hatake.app.tab.0.close')), findsNothing);
  });

  testWidgets('アプリの上書きが定義より強い（タブと書いてあっても遷移で使える）',
      (tester) async {
    await tester.pumpWidget(_harness(
      _app(navigation: AppNavigation.tabs),
      navigation: AppNavigation.single,
    ));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNothing);
  });

  testWidgets('逆向きも効く（定義は既定のまま、アプリがタブにする）', (tester) async {
    await tester.pumpWidget(_harness(_app(), navigation: AppNavigation.tabs));
    await tester.pumpAndSettle();
    expect(find.byType(InputChip), findsNWidgets(1));
  });
}

File: typescript/test/appTabs.test.ts
import { describe, expect, it } from "vitest";
import { parse as parseYaml } from "yaml";
import {
  diffDefinitions,
  explainSource,
  findWarnings,
  MAX_TABS,
  parseAppYaml,
  parsePageYaml,
} from "../src/index.js";

/// 画面を**並べて開く**（`app.navigation`）と、**遷移のボタンがどこに開くか**
/// （`action.open`）。
///
/// 業務システムによって作法が違う（1件ずつ処理する伝票と、見比べながら直すマスタ）ので、
/// どちらかに決め打ちしない。定義が既定を言い、アプリ側が上書きする。
///
/// 道具の側で言うことは3つ: **効かない所を言う**（並べないアプリの `open: tab`・遷移
/// ではないボタン・知らない値）・**読み返す**（どう開くか／このボタンは別のタブか）・
/// **変わったら言う**（開き方が変わるのは使い方が変わること）。
type Dict = Record<string, unknown>;

const app = (extra: string, actionExtra = "") => `
app:
  id: shop
  title: 店
${extra}  menu:
    - { id: c, label: 顧客, page: customers }
  pages:
    - type: search
      id: customers
      title: 顧客一覧
      repository: repo
      table:
        columns: [{ field: code, label: コード }]
        rowActions: [open]
      actions:
        - id: open
          type: navigate
          label: 詳細
          page: customer_detail
${actionExtra}
    - type: detail
      id: customer_detail
      title: 顧客詳細
      repository: repo
      form: { sections: [{ fields: [{ field: code, label: コード }] }] }
`;

const TABS = "  navigation: tabs\n";
const OPEN_TAB = "          open: tab";

const rules = (source: string): string[] =>
  findWarnings(parseYaml(source) as Dict).map((one) => one.rule);

describe("解析", () => {
  it("書かなければ 1画面ずつ（後方互換）", () => {
    const parsed = parseAppYaml(app(""), { strict: true });
    expect(parsed.navigation).toBe("single");
    expect(parsed.pages[0].type).toBe("search");
  });

  it("並べて開くと書ける（strict でも通る）", () => {
    expect(parseAppYaml(app(TABS), { strict: true }).navigation).toBe("tabs");
  });

  it("遷移のボタンは、どこに開くかを1件ずつ書ける（既定は同じ画面の続き）", () => {
    const page = parsePageYaml(
      `page:
  type: search
  id: customers
  title: 顧客一覧
  repository: repo
  table:
    columns: [{ field: code, label: コード }]
  actions:
    - { id: a, type: navigate, label: 詳細, page: d }
    - { id: b, type: navigate, label: 別タブ, page: d, open: tab }
`,
      { strict: true },
    );
    expect(page.actions[0].open).toBe("same");
    expect(page.actions[1].open).toBe("tab");
  });
});

describe("効かない所を言う", () => {
  it("並べて開くアプリなら黙る", () => {
    expect(rules(app(TABS, OPEN_TAB))).toEqual([]);
  });

  it("並べないアプリの `open: tab` は効かない", () => {
    const found = findWarnings(parseYaml(app("", OPEN_TAB)) as Dict);
    expect(found.map((one) => one.rule)).toEqual(["open-without-tabs"]);
    expect(found[0].path).toBe("app.pages[0].actions[0].open");
    expect(found[0].message).toContain("同じ画面の続き");
    // アプリ側で上書きしている場合もあるので、そう書いておく（嘘の断定をしない）。
    expect(found[0].fix).toContain("アプリ側で上書き");
  });

  it("遷移ではないボタンに書いても、開く先が無い", () => {
    const found = findWarnings(
      parseYaml(`page:
  type: search
  id: customers
  title: 顧客一覧
  repository: repo
  table:
    columns: [{ field: code, label: コード }]
  actions:
    - { id: csv, type: export, label: CSV, open: tab }
`) as Dict,
    );
    expect(found.map((one) => one.rule)).toEqual(["open-without-navigate"]);
  });

  it("知らない値は黙って既定になるので、そう言う", () => {
    expect(rules(app("  navigation: split\n"))).toContain("unknown-navigation");
    expect(rules(app(TABS, "          open: window"))).toContain("unknown-open");
  });
});

describe("読み返す", () => {
  const opening = (source: string): string[] =>
    explainSource(source).sections.find((one) => one.title === "画面の開き方")
      ?.lines ?? [];

  it("どう開くかを言う（上限も・上書きできることも）", () => {
    const tabs = opening(app(TABS));
    expect(tabs[0]).toContain("並べて開く");
    expect(tabs[0]).toContain(`最大 ${MAX_TABS} 枚`);
    expect(tabs[1]).toContain("アプリ側で上書きできます");

    expect(opening(app(""))[0]).toContain("1画面ずつ開く");
  });

  it("別のタブで開くボタンは、そう読み返す", () => {
    const actions = explainSource(app(TABS, OPEN_TAB), { page: "customers" })
      .sections.find((one) => one.title === "できる操作");
    expect(actions?.lines.join("\n")).toContain("別のタブで開く");
    // 既定（same）のボタンには何も足さない（当たり前のことを言わない）。
    const plain = explainSource(app(TABS), { page: "customers" }).sections.find(
      (one) => one.title === "できる操作",
    );
    expect(plain?.lines.join("\n")).not.toContain("別のタブ");
  });
});

describe("変わったら言う", () => {
  it("開き方が変わるのは、使い方が変わること", () => {
    const { changes } = diffDefinitions(
      parseYaml(app("")) as Dict,
      parseYaml(app(TABS)) as Dict,
    );
    const one = changes.find((c) => c.kind === "navigation-changed");
    expect(one?.impact).toBe("caution");
    expect(one?.message).toContain("並べて開く");
    expect(one?.message).toContain("同じ画面は2枚開きません");
  });

  it("ボタンの開き方も言う（増えるだけなので safe）", () => {
    const { changes } = diffDefinitions(
      parseYaml(app(TABS)) as Dict,
      parseYaml(app(TABS, OPEN_TAB)) as Dict,
    );
    const one = changes.find((c) => c.kind === "open-changed");
    expect(one?.impact).toBe("safe");
    expect(one?.message).toContain("別のタブで開く");
  });
});
